### PR TITLE
make RepoConfig and CommandBase use absolute system paths

### PR DESCRIPTION
**Pull request by arlyon**
### Description

Continue the correctness work by making the base command provide an AbsoluteSystemPathBuf

### CODE REVIEW

File: crates/turborepo-lib/src/cli.rs [hunk 1]
```diff
@@ -488,6 +489,9 @@ pub async fn run(repo_state: Option<RepoState>) -> Result<Payload> {
         current_dir()?
     };
 
+    // a non-absolute repo root is a bug
```
Comment (arlyon) [inline crates/turborepo-lib/src/cli.rs:492]: If the lines above do not give an absolute path it is programmer error.

**Review by NicholasLYang — APPROVED** [review]

File: crates/turborepo-lib/src/commands/link.rs [hunk 2]
```diff
@@ -166,8 +167,11 @@ pub async fn link(
             verify_caching_enabled(&api_client, team_id, token, Some(selected_team.clone()))
                 .await?;
 
-            fs::create_dir_all(base.repo_root.join(".turbo"))
-                .context("could not create .turbo directory")?;
+            fs::create_dir_all(
+                base.repo_root
+                    .join_relative(RelativeSystemPathBuf::new(".turbo").expect("relative")),
```
[the remark this answers is not in the document]
Comment (chris-olszewski) [reply crates/turborepo-lib/src/commands/link.rs:172]: I think we could pretty much copy this example from once_cell: https://github.com/matklad/once_cell/blob/master/examples/regex.rs#L5

File: crates/turborepo-lib/src/commands/mod.rs [hunk 3]
```diff
@@ -158,19 +161,17 @@ impl CommandBase {
 #[cfg(test)]
 mod test {
     use test_case::test_case;
+    use turbopath::AbsoluteSystemPathBuf;
 
     use crate::get_version;
 
     #[test_case("/tmp/turborepo", "6e0cfa616f75a61c"; "basic example")]
```
Comment (chris-olszewski) [inline crates/turborepo-lib/src/commands/mod.rs:168]: Looks like windows isn't happy with `/tmp/turborepo`

**Review by chris-olszewski — APPROVED** [review]

File: crates/turborepo-lib/src/commands/link.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (arlyon) [reply crates/turborepo-lib/src/commands/link.rs:172]: I will do that now. I started a 'common_paths' module but decided against it. That macro looks ideal.

[the remark this answers is not in the document]
Comment (arlyon) [reply crates/turborepo-lib/src/commands/link.rs:172]: I think we could do a proc macro that checks the syntax at build time.

**Pull request merged**